sklearn/preprocessing/_function_transformer.py

```diff
@@ -241,29 +269,53 @@ def transform(self, X):
         X = self._check_input(X, reset=False)
         out = self._transform(X, func=self.func, kw_args=self.kw_args)
+        output_config = _get_output_config("transform", self)["dense"]
 
         if hasattr(out, "columns") and self.feature_names_out is not None:
-            # check the consistency between the column names of the output and the
-            # one generated by `get_feature_names_out`
-            if list(out.columns) != list(self.get_feature_names_out()):
-                raise ValueError(
-                    "The output generated by `func` have different column names than "
-                    "the one generated by the method `get_feature_names_out`. "
-                    f"Got output with columns names: {list(out.columns)} and "
-                    "`get_feature_names_out` returned: "
-                    f"{list(self.get_feature_names_out())}. "
-                    "This can be fixed in different manners depending on your use case:"
-                    "\n(i) If `func` returns a container with column names, make sure "
-                    "they are consistent with the output of `get_feature_names_out`.\n"
-                    "(ii) If `func` is a NumPy `ufunc`, then forcing `validate=True` "
-                    "could be considered to internally convert the input container to "
-                    "a NumPy array before calling the `ufunc`.\n"
-                    "(iii) The column names can be overriden by setting "
-                    "`set_output(transform='pandas')` such that the column names are "
-                    "set to the names provided by `get_feature_names_out`."
+            # check the consistency between the column provided by `transform` and
+            # the the column names provided by `get_feature_names_out`.
+            feature_names_out = self.get_feature_names_out()
+            if list(out.columns) != list(feature_names_out):
+                # we can override the column names of the output if it is inconsistent
+                # with the column names provided by `get_feature_names_out` in the
+                # following cases:
+                # * `func` preserved the column names between the input and the output
+                # * the input column names are all numbers
+                # * the output is requested to be a DataFrame (pandas or polars)
+                feature_names_in = getattr(
+                    X, "feature_names_in_", _get_feature_names(X)
                 )
+                same_feature_names_in_out = feature_names_in is not None and list(
+                    feature_names_in
+                ) == list(out.columns)
+                not_all_str_columns = not all(
+                    isinstance(col, str) for col in out.columns
+                )
+                if (
+                    same_feature_names_in_out
+                    or not_all_str_columns
+                    or _is_registered_adapter(adapter_name=output_config)
+                ):
```

Comment:
I'm trying to figure out why `_is_registered_adapter` is required. It looks like all the test pass with this:

```suggestion
                if same_feature_names_in_out or not_all_str_columns:
```

Reply:
This is looking at `output_config`, the configuration for `FunctionTransformer`, which is already validated in `_get_output_config`.

If `func` requires something we do not recognize, then `_get_adapter_from_container` will raise an error.

Reply:
You are right we can delegate indeed.